src/psyclone/psyir/nodes/array_mixin.py

```diff
@@ -598,6 +598,121 @@ def get_outer_range_index(self):
                 return child.position
         raise IndexError
 
+    def same_range(self, index: int, array2, index2: int) -> bool:
+        ''' This method compares the range length of this array node
+        at a given index with the range length of a second array at
+        a second index. This is useful to verify if array operations
+        are valid, e.g.: A(3,:,5) + B(:,2,2).
+
+        Note that this check supports symbolic comparisons, e.g.:
```

Comment:
Without looking at the actual transformation, but shouldn't this compare `upper-lower` of the two ranges and the `step` (as a simplified approach ... or potentially `(upper-lower)/step`, though that likely needs a round to integer somewhere :) )?

That should allow us to compare the ranges in `a(i:i+2)` with `b(1:3)`. I am ok to leave this if this is a restriction of the transform that we don't want to change in a refactoring PR, but maybe add this as an example to the ticket #2485 (or document this, looking at the docs of the arrayrange2loop this seems not to be mentioned.

Reply:
Good point, it currently can not be done because the arrayrange2loop transformation uses a single index variable for both ranges, if the array declaration lbound is different then we need to add another expression on one of them to account for the lbound difference between both range expressions.

I will look at this when refactoring the arrayrange2loop, so I added a TODO for that one instead.

This is also similar to what @arporter is doing for WHERE cannonicalisation, so at some point we need to see if there are common functionality to re-use.

Reply:
Great, thanks!